File: tests/csharp/secondary_base_methods.cs
using System;
using Secondary.Base.Methods;

internal static class Driver {
    private static int _failed;

    private static void Check(bool ok, string msg) {
        Console.WriteLine((ok ? "  PASS " : "  FAIL ") + msg);
        if (!ok) _failed++;
    }

    public static int Main() {
        TestPrimaryBaseMethodsVisible();
        TestSecondaryBaseMethodsVisible();
        TestSecondaryBasePointerOffset();
        TestDeepChainToBaseA();
        TestDeepChainToBaseB();
        TestDeepChainToBaseC();
        TestDeepChainToBaseD();
        TestDiamondInheritance();

        Console.WriteLine();
        if (_failed == 0) {
            Console.WriteLine("All secondary_base_methods tests passed.");
            return 0;
        }
        Console.WriteLine($"{_failed} test(s) failed.");
        return 1;
    }

    // The derived class should expose methods from its primary base via the
    // usual C# inheritance.
    private static void TestPrimaryBaseMethodsVisible() {
        Console.WriteLine("TestPrimaryBaseMethodsVisible: primary base methods on derived");
        using var node = new DerivedTextNode();
        node.set_id(42);
        Check(node.get_id() == 42, $"get_id (got: {node.get_id()})");
    }

    // C# has no multiple inheritance, so methods from a secondary C++ base
    // must be re-exposed on the derived class (otherwise callers lose access
    // to them entirely).  This test verifies that re-exposure happens.
    private static void TestSecondaryBaseMethodsVisible() {
        Console.WriteLine("TestSecondaryBaseMethodsVisible: secondary base methods on derived");
        using var node = new DerivedTextNode();
        node.set_text("hello");
        Check(node.get_text() == "hello", $"get_text (got: '{node.get_text()}')");
    }

    // When a method takes a pointer to a secondary base, the C# binding must
    // pass the correctly-offset pointer. This test verifies the pointer offset
    // is applied (the critical bug fix for multiple inheritance).
    private static void TestSecondaryBasePointerOffset() {
        Console.WriteLine("TestSecondaryBasePointerOffset: passing derived as secondary base");
        using var container = new Container();
        using var node = new DerivedTextNode();
        node.set_id(99);

        // Container::attach_child takes PrimaryBase*, which is the secondary base
        // of DerivedTextNode (TextLike is primary). The binding must apply the
        // pointer offset when passing node to attach_child.
        container.attach_child(node);

        int lastId = container.get_last_child_id();
        Check(lastId == 99, $"Container received correct pointer (got ID: {lastId}, expected: 99)");
    }

    // Deep chain tests - these test recursive upcast chain finding
    // Level4 -> Level3 -> Level2 -> Level1 -> BaseA (4 hops!)
    private static void TestDeepChainToBaseA() {
        Console.WriteLine("TestDeepChainToBaseA: 4-hop chain Level4->Level3->Level2->Level1->BaseA");
        using var obj = new Level4();
        obj.set_a(111);
        obj.set_id(42);  // Also test that PrimaryBase is accessible

        using var tester = new DeepTester();
        tester.test_base_a(obj);

        Check(tester.get_last_a() == 111, $"BaseA via 4-hop chain (got: {tester.get_last_a()}, expected: 111)");
        Check(obj.get_id() == 42, $"PrimaryBase methods work (got: {obj.get_id()}, expected: 42)");
    }

    // Level4 -> Level3 -> Level2 -> BaseB (3 hops)
    private static void TestDeepChainToBaseB() {
        Console.WriteLine("TestDeepChainToBaseB: 3-hop chain Level4->Level3->Level2->BaseB");
        using var obj = new Level4();
        obj.set_b(222);

        using var tester = new DeepTester();
        tester.test_base_b(obj);

        Check(tester.get_last_b() == 222, $"BaseB via 3-hop chain (got: {tester.get_last_b()}, expected: 222)");
    }

    // Level4 -> Level3 -> BaseC (2 hops)
    private static void TestDeepChainToBaseC() {
        Console.WriteLine("TestDeepChainToBaseC: 2-hop chain Level4->Level3->BaseC");
        using var obj = new Level4();
        obj.set_c(333);

        using var tester = new DeepTester();
        tester.test_base_c(obj);

        Check(tester.get_last_c() == 333, $"BaseC via 2-hop chain (got: {tester.get_last_c()}, expected: 333)");
    }

    // Level4 -> BaseD (1 hop, direct secondary base)
    private static void TestDeepChainToBaseD() {
        Console.WriteLine("TestDeepChainToBaseD: 1-hop chain Level4->BaseD");
        using var obj = new Level4();
        obj.set_d(444);

        using var tester = new DeepTester();
        tester.test_base_d(obj);

        Check(tester.get_last_d() == 444, $"BaseD via 1-hop chain (got: {tester.get_last_d()}, expected: 444)");
    }

    // Diamond inheritance: DiamondBottom -> DiamondLeft (primary), DiamondRight (secondary)
    // Both DiamondLeft and DiamondRight inherit from DiamondBase (two separate subobjects).
    // Passing DiamondBottom as DiamondRight* must use the correct offset.
    private static void TestDiamondInheritance() {
        Console.WriteLine("TestDiamondInheritance: diamond pattern with secondary base offset");
        using var obj = new DiamondBottom();
        obj.set_left(55);
        obj.set_right(77);
        obj.set_bottom(99);

        using var tester = new DiamondTester();

        // Passing as DiamondLeft* (primary base, offset 0) should work trivially
        tester.test_left(obj);
        Check(tester.get_last_left() == 55, $"DiamondLeft via primary base (got: {tester.get_last_left()}, expected: 55)");

        // Passing as DiamondRight* (secondary base, non-zero offset) - this is the real test
        tester.test_right(obj);
        Check(tester.get_last_right() == 77, $"DiamondRight via secondary base (got: {tester.get_last_right()}, expected: 77)");
    }
}

File: tests/csharp/secondary_base_methods.cxx
#include "secondary_base_methods.h"

PrimaryBase::PrimaryBase() : _id(0) {}
PrimaryBase::~PrimaryBase() {}
int PrimaryBase::get_id() const { return _id; }
void PrimaryBase::set_id(int id) { _id = id; }

TextLike::TextLike() {}
TextLike::~TextLike() {}
void TextLike::set_text(const std::string &text) { _text = text; }
std::string TextLike::get_text() const { return _text; }

DerivedTextNode::DerivedTextNode() {}

Container::Container() : _last_child(nullptr) {}

void Container::attach_child(PrimaryBase *child) {
  _last_child = child;
}

int Container::get_last_child_id() const {
  return _last_child ? _last_child->get_id() : -1;
}

// Deep chain implementations
BaseA::BaseA() : _a(0) {}
void BaseA::set_a(int val) { _a = val; }
int BaseA::get_a() const { return _a; }

BaseB::BaseB() : _b(0) {}
void BaseB::set_b(int val) { _b = val; }
int BaseB::get_b() const { return _b; }

BaseC::BaseC() : _c(0) {}
void BaseC::set_c(int val) { _c = val; }
int BaseC::get_c() const { return _c; }

BaseD::BaseD() : _d(0) {}
void BaseD::set_d(int val) { _d = val; }
int BaseD::get_d() const { return _d; }

Level1::Level1() {}
Level2::Level2() {}
Level3::Level3() {}
Level4::Level4() {}

DeepTester::DeepTester() : _last_a(0), _last_b(0), _last_c(0), _last_d(0) {}
void DeepTester::test_base_a(BaseA *obj) { _last_a = obj->get_a(); }
void DeepTester::test_base_b(BaseB *obj) { _last_b = obj->get_b(); }
void DeepTester::test_base_c(BaseC *obj) { _last_c = obj->get_c(); }
void DeepTester::test_base_d(BaseD *obj) { _last_d = obj->get_d(); }
int DeepTester::get_last_a() const { return _last_a; }
int DeepTester::get_last_b() const { return _last_b; }
int DeepTester::get_last_c() const { return _last_c; }
int DeepTester::get_last_d() const { return _last_d; }

// Diamond inheritance implementations
DiamondBase::DiamondBase() : _diamond_val(0) {}
void DiamondBase::set_diamond_val(int val) { _diamond_val = val; }
int DiamondBase::get_diamond_val() const { return _diamond_val; }

DiamondLeft::DiamondLeft() : _left(0) {}
void DiamondLeft::set_left(int val) { _left = val; }
int DiamondLeft::get_left() const { return _left; }

DiamondRight::DiamondRight() : _right(0) {}
void DiamondRight::set_right(int val) { _right = val; }
int DiamondRight::get_right() const { return _right; }

DiamondBottom::DiamondBottom() : _bottom(0) {}
void DiamondBottom::set_bottom(int val) { _bottom = val; }
int DiamondBottom::get_bottom() const { return _bottom; }

DiamondTester::DiamondTester() : _last_left(0), _last_right(0) {}
void DiamondTester::test_left(DiamondLeft *obj) { _last_left = obj->get_left(); }
void DiamondTester::test_right(DiamondRight *obj) { _last_right = obj->get_right(); }
int DiamondTester::get_last_left() const { return _last_left; }
int DiamondTester::get_last_right() const { return _last_right; }
